cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -2303,10 +2303,14 @@ Result<FileInfo> S3FileSystem::GetFileInfo(const std::string& s) {
     auto outcome = impl_->client_->HeadBucket(req);
     if (!outcome.IsSuccess()) {
       if (!IsNotFound(outcome.GetError())) {
-        return ErrorToStatus(
-            std::forward_as_tuple("When getting information for bucket '", path.bucket,
-                                  "': "),
-            "HeadBucket", outcome.GetError());
+        auto msg = "When getting information for bucket '" + path.bucket + "': ";
+        // Bucket exists, but failed to call HeadBucket, perhaps wrong region?
+        ARROW_ASSIGN_OR_RAISE(auto region, impl_->client_->GetBucketRegion(path.bucket));
```

Comment:
This will issue an additional API call. I'm not sure it's desirable to do this ourselves, rather than simply enhance the original error message with a clue as to the possible cause?

For example:
> When getting information for bucket 'XXX' (perhaps you configured the wrong region?): ...

Reply:
Another possibility would be to condition this on the HTTP status code, which seems to be 301 in case the bucket exists in another region.

Reply:
Note that we only do this extra API call if there is an error (so it doesn't impact code that doesn't error), and also only if the bucket exists (because if it doesn't exist (eg a typo), we return FileType::NotFound). So that might already limit the cases where this extra call would actually be a problem.

Another option could be to only check for this is the AWS Error is UNKNOWN (since it doesn't seem to have a specific error code for "buckets exists but not on this region")

Reply:
Also, perhaps you can try inspecting the headers of the HeadBucket error response? The answer might be there already.

Reply:
Last thing: ideally we do not do this only for HeadBucket, but for all similar operations (such as HeadObject).

Reply:
Ah, it was in the headers after all. :+1: https://github.com/apache/arrow/pull/14601/commits/eccb0b521b7281c6ec09d9793ab2cee3137130bd